aleo-programs/staking_lite/src/main.leo

```diff
@@ -0,0 +1,376 @@
+import credits.aleo;
+
+program staking_lite.aleo {
+  // Owner of the program
+  const ADMIN: address = aleo1kf3dgrz9lqyklz8kqfy0hpxxyt78qfuzshuhccl02a5x43x6nqpsaapqru;
+  // Address of this program
+  const CORE_PROTOCOL: address = aleo17hwvp7fl5da40hd29heasjjm537uqce489hhuc3lwhxfm0njucpq0rvfny;
+  const PRECISION_UNSIGNED: u128 = 1000u128;
+  const UNBONDING_PERIOD: u32 = 360u32;
+
+  /** Commission rate: 0u8 -> u128
+    * percentage of rewards taken as commission
+    * relative to precision of 1000
+    * e.g. 100u128 = 10%
+  */
+  mapping commission_percent: u8 => u128;
+
+  // 0u8 -> address of validator
+  mapping validator: u8 => address;
+
+  /** Program controls: 0u8 ->
+    * 0u8: functioning as normal
+    * 1u8: full unbond in progress
+    * 2u8: protocol fully unbonded
+  */
+  mapping state: u8 => u8;
+
+  // 0u8 -> total pool of delegator tokens
+  mapping total_shares: u8 => u64;
+
+  // address -> number of shares held by the delegator with this address
+  mapping delegator_shares: address => u64;
+
+  // 0u8 -> total balance in microcredits, updated for each deposit and withdrawal
+  mapping total_balance: u8 => u64;
+
+  // 0u8 -> balance pending withdrawal in microcredits
+  mapping pending_withdrawal: u8 => u64;
+
+  /** Unbonding allowed: 0u8 ->
+    * The height at which the current withdrawal batch will be done unbonding
+    * if not present or == 0u32, a new batch can begin unbonding
+  */
+  mapping current_batch_height: u8 => u32;
+
+  struct withdrawal_state {
+    microcredits: u64;
+    claim_block: u32;
+  }
+
+  // address -> pending withdrawal for the delegator with this address
+  mapping withdrawals: address => withdrawal_state;
+
+  transition initialize(commission_rate: u128, validator_address: address) {
+    assert_eq(self.caller, ADMIN);
+    assert(commission_rate < PRECISION_UNSIGNED);
+
+    return then finalize(commission_rate, validator_address);
+  }
+
+  finalize initialize(commission_rate: u128, validator_address: address) {
+    commission_percent.set(0u8, commission_rate);
+    validator.set(0u8, validator_address);
+    state.set(0u8, 0u8);
+    total_shares.set(0u8, 0u64);
+    total_balance.set(0u8, 0u64);
+    pending_withdrawal.set(0u8, 0u64);
+    current_batch_height.set(0u8, 0u32);
+  }
+
+  transition initial_deposit(
```

Comment:
It seems like we should group this transition with the `initialize` transition. Is there any reason to separate them?

Reply:
There is a case where all shareholders may pull all funds out of the staking contract. Separating initial deposit would allow the validator to kickstart the liquid staking contract again. I don't know if that is a necessary feature, but that is an advantage of keeping these transitions separate.

Reply:
I think that's fair. I'm adding an additional check to ensure pending withdrawals are 9 but otherwise it should be safe. Is there anything else to check?